File: 022_generate_parenthesis.rb
=begin
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
=end

# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
    solutions = []
    backtrack('', n, 0, 0, solutions)
    
    return solutions
end

def backtrack(sol, n, left, right, solutions)
    if sol.length == 2*n
        solutions << sol
    end
    if left < n
      backtrack(sol + '(', n, left+1, right, solutions)
    end
    if right < left
        backtrack(sol + ')', n, left, right+1, solutions)
    end
end

puts generate_parenthesis(3).inspect

File: 022_generate_parenthesis.py
'''
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
'''

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        left = right = n
        sol = []
        self.backtrack(sol, left, right, n, "")
        
        return sol
    
    def backtrack(self, sol, left, right, n, combo):
        if len(combo) == 2*n:
            sol.append(combo)
            return
        if left > 0:
            self.backtrack(sol, left - 1, right, n, combo + '(')
        if right > left:
            self.backtrack(sol, left, right - 1, n, combo + ')')
        
